### ISSUE TITLE
Using mixed precision causes incorrect loss, metric values

### ISSUE THREAD

**Issue by Crispy13** [tensorflow/tensorflow#45568]
Tensorflow version : 2.4.0-rc3, compiled from source
GPU : RTX 3080 10GB
CUDA / CUDNN : 11.1 / 8
bazel version : 3.1.0
Windows 10

I decided to mixed precision to speed up the training, but some issues were.
I used this code:
```
policy = keras.mixed_precision.Policy('mixed_float16')
keras.mixed_precision.set_global_policy(policy)

...

model.compile(optimizer = keras.mixed_precision.LossScaleOptimizer(optimizers.Adam(learning_rate=1e-2))
                        loss = dice_loss
                        metric = [dice_coefficient]
                                                )
```
**Training didn't show any nan loss and metric. But when evaluating the model or each epoch ended, val loss and metric were incorrect(they were fixed over epochs).** The colab notebook is **[here](https://colab.research.google.com/drive/1kLpn88HSn7WOnH5vN5l-GlLRTPVArsDw?usp=sharing)**

etc:
- The values of `model.predict(x)` and `model(x)` are different. When the former was nan, the latter seemed normal.

**Comment by amahendrakar** [comment]
On running the code with [TF v2.2](https://colab.research.google.com/gist/amahendrakar/f32019942de7a6cba160e8ffb5b2dfd1/45568-2-2.ipynb#scrollTo=kM-cUF1sR-_k), the `val_loss` value is `1.0001`.

Was able to reproduce the issue with [TF v2.3](https://colab.research.google.com/gist/amahendrakar/25afe63903c66dc02a2a1a90b6482ca3/45568.ipynb#scrollTo=kM-cUF1sR-_k), with which the `val_loss` is `nan`.

Whereas with [TF-nightly](https://colab.research.google.com/gist/amahendrakar/e3aa6bab29586bb0449e56d48ebeeb27/45568-tf-nightly.ipynb#scrollTo=kM-cUF1sR-_k), the code runs indefinitely. Please check the linked gist for reference. Thanks!

**Comment by mhs-achyut** [comment]
same issue here. Please let me know once this is resolved.

**Comment by mhs-achyut** [comment]
@Crispy13 I have figured out the bug. You need to explicitly cast the softmax Dense layer to dtype=tf.float32 to avoid numerical instability. Thanks to the source here: https://wandb.ai/site/articles/mixed-precision-training-with-tf-keras

**Comment by saikumarchalla** [comment]
As @amahendrakar  mentioned above, the code runs indefinitely in TF 2.5. Please check the linked gist for [reference](https://colab.research.google.com/gist/saikumarchalla/b125e6a6224324e84a5385dc2e07d3d7/45568-tf-nightly.ipynb#scrollTo=kM-cUF1sR-_k&uniqifier=1). Thanks!

**Comment by PCdurham** [comment]
I have the same issue:
TF 2.5
RTX A5000 GPU
Ubuntu 20.04

I'm training a Keras implementation of Tiramisu, when using a Dice loss, I have to use float 32 in all inputs and outputs and I cannot use float16 mixed precision training. I tried things like lower lr and gradient normalisation.  Also tried the solution above of casting my softmax to float32.  But as soon as I go to float16 for the inputs , training loss and metric go to nan.  It's not a fatal error. But not using the tensor cores slows down the training by a factor of almost 2.

Is there a solution to this?

**Comment by TarasMLAI** [comment]
Hi, 
same issues, I'm following the guide [https://www.tensorflow.org/guide/mixed_precision] using  standard tf.keras.Model.fit() method, the output layer has has dtype  float32 `outputs = layers.Activation("linear", dtype=tf.float32)(outputs)`. The model is even converging (based on accuracy), significant speed up, but loss value is **nan**. Any help is appreciated.

Tensorflow 2.8.0 and 2.9.1

Epoch 1/10
31/31 - 33s - loss: nan - sparse_categorical_accuracy: 0.1689 - val_loss: nan - val_sparse_categorical_accuracy: 0.3802 - 33s/epoch - 1s/step
Epoch 2/10
31/31 - 19s - loss: nan - sparse_categorical_accuracy: 0.4168 - val_loss: nan - val_sparse_categorical_accuracy: 0.3854 - 19s/epoch - 610ms/step

**Comment by TheOldJonny** [comment]
I am facing a similar issue with Tensorflow 2.8.  I report here a toy example to reproduce the issue.  When I use mixed precision, during the training both the loss and the accuracy always remain the same. If I remove mixed precision, everything works fine.

```
import tensorflow as tf
from tensorflow.keras.datasets import mnist

gpus = tf.config.list_physical_devices('GPU')
if gpus:
  try:
    tf.config.experimental.set_memory_growth(gpus[0], True)
    logical_gpus = tf.config.list_logical_devices('GPU')
    print(len(gpus), "Physical GPUs,", len(logical_gpus), "Logical GPUs")
    print("\n\n")
  except RuntimeError as e:
    print(e) 
    
policy = tf.keras.mixed_precision.Policy('mixed_float16')
tf.keras.mixed_precision.set_global_policy(policy)

print('Compute dtype: %s' % policy.compute_dtype)
print('Variable dtype: %s' % policy.variable_dtype)

(trainX, trainy), (testX, testy) = mnist.load_data()

trainX = trainX.astype('float16') / 255
testX = testX.astype('float16') / 255



inputs = tf.keras.Input(shape=(28, 28, 1), name='digits')
conv_1 = tf.keras.layers.Conv2D(32, (3, 3), activation='relu', name = 'conv_1')
x = conv_1(inputs)
maxPool = tf.keras.layers.MaxPooling2D((2, 2), name = 'MaxPool_1')
x= maxPool(x)
conv_2 = tf.keras.layers.Conv2D(64, (3, 3), activation='relu', name = 'conv_2')
x = conv_2(x)
globMaxPool = tf.keras.layers.GlobalMaxPooling2D()
x = globMaxPool(x)
x = tf.keras.layers.Dense(10)(x)
print('outpout from the Dense layer', x.dtype.name)
outputs = tf.keras.layers.Activation(activation="softmax", dtype = 'float32')(x)
model = tf.keras.Model(inputs, outputs)

print('Outputs dtype: %s' % outputs.dtype.name)
print('Inputs dtype: %s' % inputs.dtype.name)

model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate = 0.001), metrics=["accuracy"], loss = "sparse_categorical_crossentropy")
initial_weights = model.get_weights()
model.fit(trainX, trainy, epochs = 10)
model.evaluate(testX, testy)

```

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
